akka-remote/src/main/scala/akka/remote/artery/ArterySubsystem.scala

```diff
@@ -67,30 +67,27 @@ private[remote] class ArterySubsystem(_system: ExtendedActorSystem, _provider: R
 
     val association =
       if (cached ne null) cached
-      else {
-        val association = getAssociation(remoteAddress)
-        association.associate()
-        recipient.cachedAssociation = association
-        association
-      }
+      else associate(remoteAddress)
 
     association.send(message, senderOption, recipient)
   }
 
-  private def getAssociation(remoteAddress: Address): Association = {
+  private def associate(remoteAddress: Address): Association = {
     val current = associations.get(remoteAddress)
     if (current ne null) current
     else {
-      val newAssociation = new Association(materializer, remoteAddress, transport)
-      val currentAssociation = associations.putIfAbsent(remoteAddress, newAssociation)
-      if (currentAssociation eq null) {
-        newAssociation
-      } else currentAssociation
+      associations.computeIfAbsent(remoteAddress, new java.util.function.Function[Address, Association] {
```

Comment:
It does not hurt for sure to add one more here. I will do merge anyway though and fix it in the next round as PR validation is rather costly :(